File: server/src/service/vs.ts
import { randomInt } from "node:crypto";

/**
 * Generate a numeric variable symbol, max 10 digits, unique against the set of
 * VS values currently in use by open sessions.
 *
 * Strategy: random 10-digit number (no leading zero) for high entropy and low
 * collision probability, retried until unique. Falls back to widening the
 * search space defensively (it never realistically exhausts at one-register scale).
 */
export function generateVs(taken: Set<string>): string {
  for (let attempt = 0; attempt < 1000; attempt++) {
    const candidate = randomVs();
    if (!taken.has(candidate)) return candidate;
  }
  throw new Error("failed to generate a unique VS after 1000 attempts");
}

function randomVs(): string {
  // 10-digit, first digit 1-9 so the string length is exactly 10.
  const first = randomInt(1, 10); // 1..9
  let rest = "";
  for (let i = 0; i < 9; i++) {
    rest += randomInt(0, 10).toString();
  }
  return `${first}${rest}`;
}

File: android/app/src/main/java/cz/qrplatba/service/Vs.kt
package cz.qrplatba.service

import java.security.SecureRandom

/**
 * Generate a numeric variable symbol, exactly 10 digits (no leading zero), unique
 * against the set of VS values currently in use by open sessions. Retried until unique.
 */
object Vs {
    private val rng = SecureRandom()

    fun generate(taken: Set<String>): String {
        repeat(1000) {
            val candidate = randomVs()
            if (!taken.contains(candidate)) return candidate
        }
        throw IllegalStateException("failed to generate a unique VS after 1000 attempts")
    }

    private fun randomVs(): String {
        val first = rng.nextInt(9) + 1 // 1..9
        val sb = StringBuilder().append(first)
        repeat(9) { sb.append(rng.nextInt(10)) }
        return sb.toString()
    }
}
